presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java

```diff
@@ -0,0 +1,163 @@
+                .collect(Collectors.toSet());
+
+        if (singletonInputs.isEmpty()) {
+            return Optional.empty();
+        }
+
+        // exclude any inputs to TRY expression. Inlining them would potentially
+        // change the semantics of those expressions
+        Set<Symbol> tryArguments = parent.getAssignments()
+                .getExpressions().stream()
+                .flatMap(e -> extractTryArguments(e).stream())
+                .collect(Collectors.toSet());
+
+        Set<Symbol> candidates = Sets.difference(singletonInputs, tryArguments);
+
+        // see if any assignments for those inputs is *not* an identity projection
+        boolean allIdentities = child.getAssignments()
+                .entrySet().stream()
+                .filter(e -> candidates.contains(e.getKey()))
+                .allMatch(e -> e.getValue() instanceof SymbolReference &&
+                        e.getKey().getName().equals(((SymbolReference) e.getValue()).getName()));
+        if (allIdentities) {
+            return Optional.empty();
+        }
+
+        // inline the expressions
+        Assignments.Builder parentAssignments = Assignments.builder();
+        for (Map.Entry<Symbol, Expression> entry : parent.getAssignments().entrySet()) {
+            Expression inlined = ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Object>()
+            {
+                @Override
+                public Expression rewriteSymbolReference(SymbolReference reference, Object context, ExpressionTreeRewriter treeRewriter)
+                {
+                    Symbol symbol = Symbol.from(reference);
+                    if (candidates.contains(symbol)) {
+                        return child.getAssignments().get(symbol);
+                    }
+
+                    return reference;
+                }
+            }, entry.getValue());
+
+            parentAssignments.put(entry.getKey(), inlined);
+        }
+
+        // derive identity assignments for the inputs of expressions that were inlined
+        Set<Symbol> inputs = child.getAssignments()
+                .entrySet().stream()
+                .filter(e -> candidates.contains(e.getKey()))
+                .map(Map.Entry::getValue)
+                .flatMap(e -> DependencyExtractor.extractAll(e).stream())
+                .collect(Collectors.toSet());
+
+        Assignments.Builder childAssignments = Assignments.builder();
+        for (Map.Entry<Symbol, Expression> assignment : child.getAssignments().entrySet()) {
+            if (!candidates.contains(assignment.getKey())) {
+                childAssignments.put(assignment.getKey(), assignment.getValue());
+            }
+        }
+        for (Symbol input : inputs) {
+            childAssignments.put(input, input.toSymbolReference());
+        }
+
+        return Optional.of(
+                new ProjectNode(
+                        parent.getId(),
+                        new ProjectNode(
```

Comment:
Legacy MergeProjections was merging completely two project nodes into one.
I think it still is desired behavior, especially in case when child projection is an identity projection, or when child projection are used only once in parent projection. 

It is common that plan may contain several intendenty projections each after another. With this rule the general output shape will remain the same.